src/libmongoc/tests/test-mongoc-sdam.c

```diff
@@ -215,6 +216,250 @@ test_sdam_cb (bson_t *test)
+static void
+sdam_json_test_ctx_init_config (json_test_ctx_t *ctx,
+                                const json_test_config_t *config)
+{
+   memset (ctx, 0, sizeof (*ctx));
+   ctx->config = config;
+   bson_init (&ctx->events);
+   ctx->acknowledged = true;
+   ctx->verbose = test_framework_getenv_bool ("MONGOC_TEST_MONITORING_VERBOSE");
+   bson_init (&ctx->lsids[0]);
+   bson_init (&ctx->lsids[1]);
+   bson_mutex_init (&ctx->mutex);
+}
+
+static void
+sdam_json_test_ctx_init_client (json_test_ctx_t *ctx, mongoc_client_t *client)
+{
+   const char *db_name;
+   const char *coll_name;
+
+   ctx->test_framework_uri = mongoc_uri_copy (client->uri);
+   db_name = bson_lookup_utf8 (ctx->config->scenario, "database_name");
+   coll_name = bson_lookup_utf8 (ctx->config->scenario, "collection_name");
+   ctx->client = client;
+   ctx->db = mongoc_client_get_database (client, db_name);
+   ctx->collection = mongoc_database_get_collection (ctx->db, coll_name);
+}
+
+static void
+sdam_json_test_ctx_cleanup (json_test_ctx_t *ctx)
+{
+   mongoc_collection_destroy (ctx->collection);
+   mongoc_database_destroy (ctx->db);
+   bson_destroy (&ctx->lsids[0]);
+   bson_destroy (&ctx->lsids[1]);
+   bson_destroy (&ctx->events);
+   mongoc_uri_destroy (ctx->test_framework_uri);
+   bson_destroy (ctx->sent_lsids[0]);
+   bson_destroy (ctx->sent_lsids[1]);
+   bson_mutex_destroy (&ctx->mutex);
+}
+
+static bool
+sdam_integration_operation_cb (json_test_ctx_t *ctx,
+                               const bson_t *test,
+                               const bson_t *operation)
+{
+   bson_t reply;
+   bool res;
+
+   res =
+      json_test_operation (ctx, test, operation, ctx->collection, NULL, &reply);
+
+   bson_destroy (&reply);
+
+   return res;
+}
+
+/* Try to get a completely clean slate by disabling failpoints on all servers.
+ */
+static void
+deactivate_failpoints_on_all_servers (mongoc_client_t *client)
+{
+   int i;
+   uint32_t server_id;
+   mongoc_set_t *servers;
+   bson_t cmd;
+   bson_error_t error;
+
+   bson_init (&cmd);
+   BCON_APPEND (&cmd, "configureFailPoint", "failCommand", "mode", "off");
+
+   servers = client->topology->description.servers;
+
+   for (i = 0; i < servers->items_len; i++) {
+      bool ret;
+
+      server_id = servers->items[i].id;
+      ret = mongoc_client_command_simple_with_server_id (client,
+                                                         "admin",
+                                                         &cmd,
+                                                         NULL /* read prefs */,
+                                                         server_id,
+                                                         NULL /* reply */,
+                                                         &error);
+      if (!ret) {
+         MONGOC_DEBUG ("error disabling failpoint: %s", error.message);
+      }
+   }
+
+   bson_destroy (&cmd);
+}
+
+static void
+run_one_integration_test (json_test_config_t *config, bson_t *test)
```

Comment:
With the introduction of threaded operations, it seemed easier to implement the test runner outside of the general `run_json_general_test`. When there is a unified test runner spec, maybe we can consolidate this.